leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/impl/QueuedRequestSender.java

```diff
@@ -114,17 +118,25 @@ public static Builder builder() {
 
         String endpoint = destination.getEndpoint();
 
-        // Accept messages only when the client is already known to ClientRegistry
-        if (clientRegistry.get(endpoint) != null) {
-            QueuedRequest queuedRequest = new QueuedRequestImpl(endpoint, castedDownlinkRequest, requestTicket);
-            messageStore.add(queuedRequest);
-            // If Client is reachable and this is the first message, we send it
-            // immediately.
-            if (clientStatusTracker.startClientReceiving(endpoint)) {
-                processingExecutor.execute(newRequestSendingTask(endpoint));
+        readWriteLock.readLock().lock();
+        try {
+            // Accept messages only when the client is already known to ClientRegistry
+            if (clientRegistry.get(endpoint) != null) {
+                QueuedRequest queuedRequest = new QueuedRequestImpl(endpoint, castedDownlinkRequest, requestTicket);
+                messageStore.add(queuedRequest);
+                // If Client is reachable and this is the first message, we send it
+                // immediately.
+                if (clientStatusTracker.startClientReceiving(endpoint)) {
+                    processingExecutor.execute(newRequestSendingTask(endpoint));
+                }
+            } else {
+                String message = String.format("message received in Queue Mode for the unknown client [%s]", endpoint);
+                LOG.warn(message);
+                // notify application layer that the message will not be sent with an exception.
+                throw new RequestCanceledException(message);
```

Comment:
The canceled notification should not be raised like other errors ? (through `ResponseListener`)